### ISSUE TITLE
Cluster health endpoint

### ISSUE THREAD

**Issue by npepinpe** [camunda/camunda#20921]
**Is your feature request related to a problem? Please describe.**

It's difficult to monitor the availability/uptime of a Zeebe cluster at the moment, as there is no explicit, cluster wide endpoint to do so. Furthermore, cluster health is poorly define, making it difficult to interpret responses from existing endpoints (e.g. Topology).

I would like to be able to monitor my Zeebe cluster's uptime from an external system, allowing me to react when it is not.

**Describe the solution you'd like**

There is a public API endpoint, e.g. `/v2/public/health`, available through the gateway, which returns 204 when the cluster is available, and 503 when it's not.

> [!Note]
> We discussed with @tmetzke to put public paths (i.e. where authentication is not enabled) under `/v2/public`. So even if we don't go with `health` as the final segment, it should be under `/v2/public`. Do verify before merging with the C8 API team however. There's already [/v2/license](https://docs.camunda.io/docs/next/apis-tools/orchestration-cluster-api-rest/specifications/get-license/) as a public endpoint.

Cluster wide health definition would be documented, and for now we can go with the existing definition: the gateway is aware of at least one partition with a healthy leader.

We can, of course, take this opportunity to come up with a better health definition.

> [!Note]
> That we are not going with the health indicators or probes framework, as these are exposed through the actuator, and here we want a _publicly_ available endpoint.

**Open questions**

- We should cope with short lived elections, as these can happen normally. So the state of "there is a healthy leader" should likely be over a window of time.
- If we have instability and lots of elections, we could have a healthy leader every 5 minutes, but it would severely impact progress. Say we have an election every minute - the system is then in a heavily degraded state, but it has a leader at some point, is the system really up?
  - imo we can ignore this for the first iteration.

**Describe alternatives you've considered**

You can currently sort of rely on the gateway readiness endpoint. Meaning if you get the topology successfully, the gateway has at least 1 known partition with a leader, and is part of the cluster. So it is able to do _some_ things, but not necessarily all. It doesn't indicate that everything is available, but rather that at least something is.

**Additional context**

- https://github.com/camunda/product-hub/issues/1755

**Comment by npepinpe** [comment]
@deepthidevaki, @lenaschoenburg - happy to start a discussion on the health definition. Note that anything below is always from the gateway point of view.

One option is what's listed above, i.e. the existing one:

- There is at least one partition with a healthy leader.

However, this means some things may be unavailable:

- If partition 1 is unhealthy, no deployments or timer start events.
- If partition X is unhealthy, then anything routed to that partition is unavailable.
- If you have 100 partitions across your cluster, and only 1 is healthy, it looks like your cluster is unavailable 99 requests out of 100 (e.g. for create process instance).

An alternative could be:

- All partitions have a healthy leader.

The downsides here:

- If 1 partition out of 100 is unhealthy, your cluster is unhealthy. But the truth is, almost all requests would be successful, so it's not really down.

**Comment by lenaschoenburg** [comment]
> If 1 partition out of 100 is unhealthy, your cluster is unhealthy. But the truth is, almost all requests would be successful, so it's not really down

How about we distinguish between healthy, _degraded_ and unhealthy? I think we need a third state because the definition of unhealthy can be either too strict or too loose, depending on the exact use case.

**Comment by npepinpe** [comment]
We can do that, but typically for uptime monitoring you want a binary status, up/down, so we would still need to advise on what is up/down (is degraded up? down?).

**Comment by lenaschoenburg** [comment]
I'd say degraded is still up.

**Comment by deepthidevaki** [comment]
> * There is at least one partition with a healthy leader.

This is a good start for the first iteration. In addition, the endpoint can include a more detailed response - how many or which partitions are healthy or unhealthy. This would help if the monitoring wants to consider degraded state later.

**Comment by npepinpe** [comment]
That's doable, but we should keep the basic uptime monitoring (based purely on status code) in mind. Though let's involved SREs there as "users" to confirm.

**Comment by multani** [comment]
@npepinpe The status should reflect whether or not the cluster is functional.
Details can be put into the payload of the response with perhaps a `status` flag expressing the actual status.

Something that can also be considered is for the endpoint to support a parameter with the expected status and change the response code based on whether or not the actual status match what the caller is expecting. For instance: `GET /v2/health?expected=healthy`

Elasticsearch is doing something similar with [`wait_for_status`](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-health.html) for instance.

**Comment by npepinpe** [comment]
I like that endpoint a lot - being able to specify what is the expected state, awaiting a specific state, etc. Being able to specify something like "wait until the topology is complete" would be pretty nice :slightly_smiling_face:

**Comment by multani** [comment]
The "wait for something" flag is not needed for the first iteration (the healthy/not healthy status is enough already) and adding it should be backward compatible, so it can definitely be added later on :+1:

**Comment by npepinpe** [comment]
I'm trying to figure out the priority at the moment. As this is part of an epic your team is driving, do you see it as a requirement for that epic?

**Comment by npepinpe** [comment]
I'm pushing it to the backlog for now. Let me know if the priority changes.

**Comment by multani** [comment]
We have rediscussed this issue for 8.8 and the applications (Operate, Tasklist, Zeebe, etc.) being merged behind a single API.

How would this endpoint work in 8.8, especially in the context of SaaS?

* Would there be a single endpoint, reporting the status of each of the underlying applications?
* Will there be a use-case where a single Camunda cluster will be deployed across several different set of deployments (for instance, one deployment dedicated for the "Zeebe" API, one deployment for the "Operate" API, similar to what we are doing < 8.8)?
  * In this case, will we need to poll each entry point separately, or will there be a single endpoint to poll the complete status of all the components?
  * If we need to poll each endpoint separately, would they all offer the same underlying API with the same guarantees?

**Comment by npepinpe** [comment]
All of these answers are assuming we don't backport this 🙃 One key fact to understand is that every component is part of the cluster from now on - even if deployed separately, they all form a single cluster.

> Would there be a single endpoint, reporting the status of each of the underlying applications?

The original issue was about the Zeebe cluster only, so we need to rewrite a bit, but the idea is to have a single endpoint reporting the health of the cluster. Would you really need the health of each application as details for your uptime monitoring?

> Will there be a use-case where a single Camunda cluster will be deployed across several different set of deployments (for instance, one deployment dedicated for the "Zeebe" API, one deployment for the "Operate" API, similar to what we are doing < 8.8)?

Yes, the plan for SaaS is to have a presentation service - UIs, APIs - that is stateless and accessible to clients, and then the brokers in their own stateful set. However, the endpoint will be on the gateway, as if the gateways are down, the whole cluster is unavailable.


> In this case, will we need to poll each entry point separately, or will there be a single endpoint to poll the complete status of all the components?
> If we need to poll each endpoint separately, would they all offer the same underlying API with the same guarantees?

Single endpoint in any of the gateways.

**Comment by npepinpe** [comment]
@mustafadagher - make sure to invite me, Lena, or Deepthi to a kick off so we can finalize the whole thing, since the issue is a little bit outdated. I don't think we need much, but just to make sure we're all on the same page 👍

**Comment by multani** [comment]
> The original issue was about the Zeebe cluster only, so we need to rewrite a bit, but the idea is to have a single endpoint reporting the health of the cluster.
> Would you really need the health of each application as details for your uptime monitoring?

I re-read the latest version of [our SLA](https://legal.camunda.com/licensing-and-other-legal-terms), and it seems now we measure all the components as a whole, so maybe we don't need to monitor each applications separately anymore.

* If we have a single monitoring endpoint 
* and if this endpoint returning `HTTP != 200` means: none of the applications are available
* and we want to account this for a "downtime event" for the whole cluster

... then I think that works for us and it could simplify our monitoring. I'll double check with PM in https://github.com/camunda/product-hub/issues/2308 to be sure 👍

**Comment by shay-matasaro-camunda** [comment]
@npepinpe how does the last comment from @multani effect the need for this fix, should we still look to work on it ? cc @mustafadagher

**Comment by npepinpe** [comment]
We still need to do this regardless, we don't have a public health endpoint with 8.8 anymore. Please schedule a kick off with @lenaschoenburg (who will support and do reviews), and @multani (or anyone from his team if he delegates) as the customer/end user.

**Comment by mustafadagher** [comment]
👋 Hello everyone,
Today, I've met with @npepinpe and @deepthidevaki to discuss this topic, we went over the current state of `uptime monitoring`, and also all the suggestions and the open questions in comments above.

**The following are the key outcomes of our discussion:**

- The definition for `UP` shall remain as `the gateway is aware of at least one partition with a healthy leader`. This is currently how **uptime** was calculated for Zeebe so far, and we do not want to change the uptime definition without agreement with other stakeholders.
- We will rely on information available from the broker topology endpoint.
- Currently the orchestration cluster also includes other webapps (**Tasklist** and **Operate**), but the current uptime definition only covers the health of **Zeebe**. 
    - This is already the current state of affairs, and it shall remain what we aim for right now.
    - The health of Tasklist/Operate will not be part of the uptime check for now. (out of scope)
    - @npepinpe will follow-up with @felix-mueller about this for further product discovery
- We suggest using different endpoint path for this use-case (`/v2/status` or `/v2/public/status`), since  the definition of `health` in our [docs](https://docs.camunda.io/docs/next/components/zeebe/technical-concepts/health/) diverges from the definition of `UP` needed for this use-case

**Comment by mustafadagher** [comment]
@npepinpe @deepthidevaki I've created a [PR](https://github.com/camunda/camunda/pull/37189) with an endpoint fulfilling the requirements we agreed upon. 

In the meantime, I've tried to investigate how the **Operate** and **Tasklist** webapps reported their health, and I found the following:
- **Operate**: [IndicesHealthIndicator](https://github.com/camunda/camunda/blob/929c78ed19e0c867b49981901505b0cb924b9e65/operate/schema/src/main/java/io/camunda/operate/management/IndicesHealthIndicator.java#L37-L44)
    - if `isHealthCheckEnabled` is enabled (true by default): Checks if the ES/OS cluster health endpoint responds with a non-RED status ([ref](https://github.com/camunda/camunda/blob/9e12eb5b38cb22a77a3b26053715375324004e22/schema-manager/src/main/java/io/camunda/search/schema/elasticsearch/ElasticsearchEngineClient.java#L308-L322))
    - AND schema manager finds all configured indices `schemaManager.isAllIndicesExist()`
- **Tasklist**: [SearchEngineHealthIndicator](https://github.com/camunda/camunda/blob/2569a5fecea11e9967b722b7adad4b0b267f2781/tasklist/els-schema/src/main/java/io/camunda/tasklist/management/SearchEngineHealthIndicator.java#L31-L38)
    -  if `isHealthCheckEnabled` is enabled (true by default): only checks that the schema exists with all indices 

The checks above are used together with the Spring Application's `readinessState` to report readiness of these apps.

💭 There's not an easy straight forward way to check the same directly from the gateway rest layer right now, something has to be done to make it available, if possible.

@npepinpe if you already managed to follow up with product about that, let me know. But we can do that separately in another issue if found necessary.

**Comment by mustafadagher** [comment]
@npepinpe Also, probably a little late to ask this question, but 
❓ Is there a reason to favor responding with `204`/`503` over always responding `200 OK` with a `status` field in the response body (to be an enum with values like `UP`, `DOWN`?

I'm asking mainly because in the review comments @deepthidevaki asked if we should implement an integration test for the new endpoint, that requires adding it to the `camunda-java-client` too. And there, dealing with response body is going to be much more easier.

**Comment by npepinpe** [comment]
It should not be required to look at the body. We can add things in the body if we think it's interesting, but it should be possible to determine coarsely if things are good or not via the status code.

**Comment by multani** [comment]
I agree with @npepinpe here. I would add that from a monitoring standpoint, being able to based the status on the HTTP code itself without to parse the body:

1. Makes monitoring the endpoint much easier/simpler: otherwise, we have to implement the parsing logic into the monitoring system, which opens a whole new can of potential problems
2. Allows to observe the status from other means: HTTP logs, HTTP proxy metrics, etc.

Feel free to add what you want in the body, as long as its consistent with the HTTP status code and that the status code gives the overall status of the system.

**Comment by mustafadagher** [comment]
@npepinpe btw, does this need to be in 8.8?

**Comment by mustafadagher** [comment]
Still needs to be backported to `stable/8.8`, blocked by #37374

**Comment by npepinpe** [comment]
@mustafadagher - can you please follow up with SREs so they can make use of this? At least a manual test with the SNAPSHOT so we can tell that it will work when comes the release.

**Comment by mustafadagher** [comment]
> [@mustafadagher](https://github.com/mustafadagher) - can you please follow up with SREs so they can make use of this? At least a manual test with the SNAPSHOT so we can tell that it will work when comes the release.

Sure. I couldn't find a slack channel for the parent epic, but I will message @multani and @hisImminence privately 👍

**Comment by mustafadagher** [comment]
> [@mustafadagher](https://github.com/mustafadagher) - can you please follow up with SREs so they can make use of this? At least a manual test with the SNAPSHOT so we can tell that it will work when comes the release.

I also brought it up in a comment on this issue https://github.com/camunda/camundaclusters-watcher/issues/255 ([here](https://github.com/camunda/camundaclusters-watcher/issues/255#issuecomment-3275389487)).
